src/back/space/src/service/service.cpp

```diff
@@ -0,0 +1,364 @@
+type: object
+description: Main service component
+additionalProperties: false
+properties:
+  can-create:
+    type: string
+    description: Can create new spaces
+  default-space:
+    type: string
+    description: Default space to join
+  spaces-limit-for-user:
+    type: string
+    description: How many spaces user can join
+  items-limit-for-list:
+    type: string
+    description: How many items list may contain
+)");
+}
+
+Service::Service(
+		const components::ComponentConfig& conf,
+		const components::ComponentContext& ctx)
+	: components::LoggableComponentBase{conf, ctx}
+	, _repo{ctx.FindComponent<Repository>()}
+	, _canCreate{conf["can-create"].As<bool>()}
+	, _defaultSpace{conf["default-space"].As<std::string>()}
+	, _spacesLimitForUser{conf["spaces-limit-for-user"].As<int>()}
+	, _itemsLimitForList{conf["items-limit-for-list"].As<int>()}
+{}
+
+PagingResult<model::Space> Service::GetList(const std::string& userId, const unsigned int start, const unsigned int limit)
+{
+	if (!_defaultSpace.empty()) {
+		const auto defSpace = _repo.Space().SelectByKey(_defaultSpace);
+		if (!_repo.SpaceUser().IsUserInside(defSpace.id, userId)) {
+			const auto nowDT = std::chrono::system_clock::now();
+			const auto now = std::chrono::duration_cast<std::chrono::seconds>(nowDT.time_since_epoch()).count();
+			// todo - what default role must be set here?
+			_repo.SpaceUser().Insert(defSpace.id, userId, false, now, Role::Type::User);
+		}
+	}
+	return _repo.Space().SelectByUserId(userId, start, limit);
+}
+
+PagingResult<model::Space> Service::GetAvailableList(const std::string& userId, const unsigned int start, const unsigned int limit)
+{
+	return _repo.Space().SelectAvailable(userId, start, limit);
+}
+
+PagingResult<model::SpaceInvitation> Service::GetInvitationList(const unsigned int start, const unsigned int limit)
+{
+	return _repo.SpaceInvitation().Select(start, limit);
+}
+
+PagingResult<model::SpaceLink> Service::GetLinkList(const unsigned int start, const unsigned int limit)
+{
+	return _repo.SpaceLink().Select(start, limit);
+}
+
+PagingResult<model::SpaceUser> Service::GetUserList(const std::string& userId, const std::string& spaceId, const unsigned int start, const unsigned int limit)
+{
+	const auto spaceUuid = utils::BoostUuidFromString(spaceId);
+
+	bool isUserInside = _repo.SpaceUser().IsUserInside(spaceUuid, userId);
+	if (!isUserInside)
+		throw errors::NotFound{};
+
+	return _repo.SpaceUser().Get(spaceUuid, start, limit);
+}
+
+bool Service::isSpaceExistsByKey(const std::string& key) {
+	return _repo.Space().IsExists(key);
+}
+
+bool Service::isCanCreate() {
+	return _canCreate;
+}
+
+int Service::CountInvitationAvailable(const std::string& currentUserId) {
+	return _repo.SpaceInvitation().GetAvailableCount(currentUserId);
+}
+
+bool Service::CheckKeyByRegex(const std::string& key) {
+	std::string regex = "[a-z0-9_]*";
+	static const std::regex rx(regex);
+	return std::regex_match(key.c_str(),rx);
```

Comment:
1. По скольку rx создаётся 1 раз, то тут получается что regex создаётся каждый раз напрасно.
2. Обычно всё что в namespace std и работает со строками, умеет принимать std::string
3. Между аргументами функций ставим пробелы

```suggestion
	static const std::regex rx("[a-z0-9_]*");
	return std::regex_match(key, rx);
```

Reply:
Готово.